### PR TITLE
Re-sharing

### PR DESCRIPTION
**Pull request by pavelkrolevets**
- Add reshare with eip1271 signature verification according to spec
- Add threshold of old operators support
- Use types from dkg-spec and align
- Add integration tests for resharing cases
- Add docker demo for resharing
- Update re-signing to spec (eip1271 sig)
- Add more integration resign tests

### CODE REVIEW

File: cli/initiator/reshare.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+			false,
+			1,
+			cli_utils.OwnerAddress,
+			cli_utils.Nonce,
+			cli_utils.WithdrawAddress,
+			cli_utils.OutputPath,
+		); err != nil {
+			logger.Fatal("Could not save results", zap.Error(err))
+		}
+		logger.Info("🚀 Resigning ceremony completed")
```
Comment (radiken) [inline cli/initiator/reshare.go:150]: Should be resharing

File: cli/initiator/resigning.go [hunk 2]
```diff
@@ -80,10 +97,23 @@ var StartResigning = &cobra.Command{
 				}
 				proofsData := wire.ConvertSignedProofsToSpec(arrayOfSignedProofs[i])
 				// Create a new ID.
-				id := crypto.NewID()
+				id := spec.NewID()
 				nonce := cli_utils.Nonce + uint64(i)
+				// Construct resign message
+				reshareMsg, err := dkgInitiator.ConstructResignMessage(
```
Comment (radiken) [inline cli/initiator/resigning.go:103]: Should be resignMsg

File: pkgs/initiator/initiator.go [hunk 3]
```diff
@@ -369,13 +428,87 @@ func (c *Initiator) StartResigning(id [24]byte, ids []uint64, proofs []*spec.Sig
...
+		Operators:     reshareMsg.SignedReshare.Reshare.NewOperators,
+		Identifier:    id,
+		DepositData:   depositData,
+		KeysharesData: keysharesData,
+		Proofs:        proofsData,
+	}
+	err = c.sendResult(resultMsg, reshareMsg.SignedReshare.Reshare.NewOperators, consts.API_RESULTS_URL, id)
 	if err != nil {
 		return nil, nil, nil, fmt.Errorf("🤖 Error storing results at operators %w", err)
 	}
```
Comment (radiken) [inline pkgs/initiator/initiator.go:514]: Can wrap these in another function as it is the same for init, resign, reshare (also commented in pull request 100)

File: pkgs/dkg/drand.go [hunk 4]
```diff
@@ -582,3 +654,258 @@ func (o *LocalOwner) Resign(reqID [24]byte, r *wire.ResignMessage) (*wire.Transp
...
+	eciesSK, pk := initsecret(o.Suite)
+	o.data.secret = eciesSK
+	bts, _, err := CreateExchange(pk, commits)
+	if err != nil {
+		return nil, err
+	}
+	return &wire.Transport{
+		Type:       wire.ReshareExchangeMessageType,
+		Identifier: reqID,
+		Data:       bts,
```
Comment (radiken) [inline pkgs/dkg/drand.go:709]: Why do we use spec.Result as the output data for Resign function above (line 639-655) but not here (and init)?

**Review by alan-ssvlabs — CHANGES_REQUESTED** [review]
We have a security report that requires discussion before approval.

File: pkgs/dkg/drand.go (hunk 4 above)
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go:650]: At reshare we have more steps, but at resign only one

File: pkgs/dkg/drand.go [hunk 5]
```diff
@@ -380,11 +388,86 @@ func (o *LocalOwner) Process(st *wire.SignedTransport) error {
...
+						return err
+					}
+				}
+			}
+			for _, op := range utils.GetDisjointNewOperators(o.data.reshare.OldOperators, o.data.reshare.NewOperators) {
+				if o.ID != op.ID {
+					continue
+				}
+				bundle := &kyber_dkg.DealBundle{}
+				b, err := wire.EncodeDealBundle(bundle)
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go:410]: we need an empty object encoded

File: pkgs/dkg/drand.go [hunk 6]
```diff
@@ -582,3 +654,258 @@ func (o *LocalOwner) Resign(reqID [24]byte, r *wire.ResignMessage) (*wire.Transp
...
+	}
+	return nil
+}
+
+func (o *LocalOwner) CheckIncomingOperators(msgs []*wire.SignedTransport) (map[uint64]*spec.Operator, error) {
+	// sanity check
+	if o.data == nil {
+		return nil, fmt.Errorf("no data object at instance")
+	}
+	// TODO: think how to change this logic to more robust
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/dkg/drand.go:782]: need to refactor this function, ill try, I dont like how it is atm

File: pkgs/initiator/initiator.go [hunk 7]
```diff
@@ -165,52 +169,119 @@ func ValidatedOperatorData(ids []uint64, operators wire.OperatorsCLI) ([]*spec.O
...
+	for _, res := range dkgResult {
+		results = append(results, res)
+	}
 	return results, nil
 }
 
+func (c *Initiator) messageFlowHandlingReshare(id [24]byte, reshareMsg *wire.ReshareMessage) ([][]byte, error) {
+	c.Logger.Info("phase 1: sending reshare message to old operators")
+	allOps := utils.JoinSets(reshareMsg.SignedReshare.Reshare.OldOperators, reshareMsg.SignedReshare.Reshare.NewOperators)
+	exchangeMsgs, errs, err := c.SendReshareMsg(id, reshareMsg, allOps)
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/initiator/initiator.go:242]: yeah, good catch, some leftovers after hundreds of iterations

File: pkgs/operator/state.go [hunk 8]
```diff
@@ -163,6 +167,96 @@ func (s *Switch) CreateResignInstance(reqID [24]byte, resign *wire.ResignMessage
...
+	var allOps []*spec.Operator
+	allOps = append(allOps, reshareMsg.SignedReshare.Reshare.OldOperators...)
+	allOps = append(allOps, reshareMsg.SignedReshare.Reshare.NewOperators...)
+	operatorID, err := spec.OperatorIDByPubKey(allOps, s.PubKeyBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+	// sanity check of operator ID
+	if s.OperatorID != operatorID {
+		return nil, nil, fmt.Errorf("wrong operator ID")
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/operator/state.go]: done at resigning, merged here

File: pkgs/operator/state.go [hunk 9]
```diff
@@ -163,6 +167,96 @@ func (s *Switch) CreateResignInstance(reqID [24]byte, resign *wire.ResignMessage
...
+	owner := dkg.New(&opts)
+	// sanity check for incoming proofs len
+	if len(reshareMsg.Proofs) != len(reshareMsg.SignedReshare.Reshare.OldOperators) {
+		return nil, nil, fmt.Errorf("wrong proofs len at reshare message: expected %d, got %d", len(reshareMsg.SignedReshare.Reshare.OldOperators), len(reshareMsg.Proofs))
+	}
+	// wait for exchange msg
+	commits, err := crypto.GetPubCommitsFromProofs(reshareMsg.SignedReshare.Reshare.OldOperators, reshareMsg.Proofs, int(reshareMsg.SignedReshare.Reshare.OldT))
+	if err != nil {
+		return nil, nil, err
+	}
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/operator/state.go:158]: I refactored here, so we have now
```
	case *wire.ReshareMessage:
		commits, share, err := s.getPublicCommitsAndSecretShare(msg)
		owner.SecretShare = share
		resp, err = owner.Reshare(reqID, &msg.SignedReshare.Reshare, commits)
		if err != nil {
			return nil, nil, err
		}
```

[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/operator/state.go:158]: probalby cant move this before `owner` creation

File: pkgs/initiator/initiator.go [hunk 10]
```diff
@@ -165,52 +169,119 @@ func ValidatedOperatorData(ids []uint64, operators wire.OperatorsCLI) ([]*spec.O
...
 	}
-	err = verifyMessageSignatures(id, results, c.VerifyMessageSignature)
+	// check that all operators replied
+	if err := checkThreshold(exchangeMsgs, errs, operators, operators, len(operators)); err != nil {
+		return nil, err
+	}
+	err = verifyMessageSignatures(id, exchangeMsgs, c.VerifyMessageSignature)
 	if err != nil {
 		return nil, err
 	}
```
[the remark this answers is not in the document]
Comment (pavelkrolevets) [reply pkgs/initiator/initiator.go:183]: do you want to move it inside? I thought it would be more readable that we check threshold each round of replies

**Pull request merged**